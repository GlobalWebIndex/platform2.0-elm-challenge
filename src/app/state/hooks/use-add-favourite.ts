import { useMutation, useQueryClient } from '@tanstack/react-query';
import { queryKeyFavourite } from './use-favourite.tsx';
import { addToFavourite } from '../../repository/favourite-helper.ts';

export const useAddFavourite = () => {
  const queryClient = useQueryClient();

  const { mutate: addFavourite, isPending } = useMutation({
    mutationFn: addToFavourite,
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: [queryKeyFavourite] });
    },
  });

  return {
    addFavourite,
    isPending,
  };
};
