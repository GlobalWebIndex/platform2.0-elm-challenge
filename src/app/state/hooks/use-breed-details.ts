import { useQuery } from '@tanstack/react-query';
import { fetchBreedDetails } from '../../repository/fetch-breed-details.ts';

export const queryKeyBreedDetails = 'breed-details';
export const useBreedDetails = (id: string) => {
  const { isPending, error, data } = useQuery({
    queryKey: [queryKeyBreedDetails, id],
    queryFn: () => fetchBreedDetails(id),
  });

  return {
    data,
    error,
    isPending,
  };
};
