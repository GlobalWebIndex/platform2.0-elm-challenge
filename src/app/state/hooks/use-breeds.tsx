import { useQuery } from '@tanstack/react-query';
import { fetchBreeds } from '../../repository/fetch-breeds.ts';

export const queryKeyBreeds = 'breeds';
export const useBreeds = () => {
  const { isPending, error, data } = useQuery({
    queryKey: [queryKeyBreeds],
    queryFn: fetchBreeds,
  });

  return {
    data,
    error,
    isPending,
  };
};
