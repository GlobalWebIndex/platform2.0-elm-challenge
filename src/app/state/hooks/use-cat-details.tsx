import { useQuery } from '@tanstack/react-query';
import { fetchCatDetails } from '../../repository/fetch-cat-details.ts';

export const queryKeyCatDetails = 'cat-details';
export const useCatDetails = (id: string) => {
  const { isPending, error, data } = useQuery({
    queryKey: [queryKeyCatDetails, id],
    queryFn: () => fetchCatDetails(id),
  });

  return {
    data,
    error,
    isPending,
  };
};
