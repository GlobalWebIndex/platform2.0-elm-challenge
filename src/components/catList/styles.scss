.container {
  display: flex;
  flex-direction: column;
  gap: 1rem;
  align-items: center;

  .cats {
    display: grid;
    gap: 1rem;
    grid-template-columns: repeat(auto-fit, minmax(200px, 1fr));
    grid-auto-rows: 200px;
    width: 100%;
  }

  .loadMore {
    $loadMore: &;

    display: flex;
    justify-content: center;
    padding-bottom: 1rem;
    width: 100%;

    button:disabled {
      @at-root #{$loadMore} {
        cursor: not-allowed;
      }
    }
  }
}

@media screen and (min-width: 768px) {
  .container {
    gap: 2rem;

    .cats {
      grid-template-columns: repeat(auto-fit, minmax(300px, 1fr));
      grid-auto-rows: 300px;
      gap: 2rem;
    }

    .loadMore {
      padding-bottom: 1.5rem;
      width: 22%;
    }
  }
}
