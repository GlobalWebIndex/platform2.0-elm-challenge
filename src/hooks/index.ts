import { useEffect, useState } from 'react';
import { useLocation } from 'react-router-dom';

export function useData<T>(
  callback: () => Promise<T>,
  handleError?: () => void
) {
  const [data, setData] = useState<T | null>(null);

  useEffect(() => {
    async function fetchData() {
      try {
        const payload = await callback.call(null);

        setData(payload);
      } catch (error) {
        handleError?.();
      }
    }

    if (data === null) fetchData();
    // eslint-disable-next-line
  }, []);

  // Need this specific code structure
  // to achieve the desired typing
  if (data === null) {
    return [false, null] as const;
  }

  return [true, data] as const;
}

export function usePageTitle() {
  const location = useLocation();
  const title = location.pathname.substring(1);
  const defaultTitle = 'My Cat App';

  return title === '' || title.includes('images') ? defaultTitle : title;
}
