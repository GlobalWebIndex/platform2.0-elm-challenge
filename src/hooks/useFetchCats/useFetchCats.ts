import {useCallback, useEffect, useState} from 'react';
import type {CatList} from 'cat-lover/types/cats';

export const useFetchCats = (): {
  isFetching: boolean;
  isReFetching: boolean;
  list: CatList;
  reFetch: () => Promise<void>;
} => {
  const [list, setList] = useState<CatList>([]);
  const [isFetching, setIsFetching] = useState(true);
  const [isReFetching, setIsReFetching] = useState(false);

  const fetchCats = useCallback<() => Promise<void>>(
    () =>
      fetch('https://api.thecatapi.com/v1/images/search?limit=10')
        .then(response => response.json() as Promise<CatList>)
        .then(data => {
          setList(prevList => [...prevList, ...data]);
        })
        .finally(() => {
          setIsFetching(false);
        }),
    []
  );

  useEffect(() => {
    void fetchCats();
  }, []);

  const reFetch = useCallback<() => Promise<void>>(async () => {
    setIsReFetching(true);

    await fetchCats();

    setIsReFetching(false);
  }, [fetchCats]);

  return {isFetching, isReFetching, list, reFetch};
};
