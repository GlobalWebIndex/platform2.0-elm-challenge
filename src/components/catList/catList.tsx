import React from 'react';
import {useLocation} from 'react-router';
import {Link} from 'react-router-dom';
import {useFetchCats} from '../../hooks/useFetchCats';
import {Loader} from '../loader';
import {Cat} from './cat';
import {LoadMoreButton} from './loadMoreButton';
import styles from './styles.scss';

export const CatList = (): JSX.Element => {
  const location = useLocation();
  const {isFetching, isReFetching, reFetch, list} = useFetchCats();

  const handleLoadMore = (): void => {
    void reFetch();
  };

  if (isFetching) {
    return <Loader />;
  }

  return (
    <div className={styles.container}>
      <div className={styles.cats}>
        {list.map((cat, index) => (
          <Link key={index} state={{backgroundLocation: location}} to={cat.id}>
            <Cat url={cat.url} />
          </Link>
        ))}
      </div>
      <div className={styles.loadMore}>
        <LoadMoreButton isDisabled={isReFetching} onLoadMore={handleLoadMore} />
      </div>
    </div>
  );
};
