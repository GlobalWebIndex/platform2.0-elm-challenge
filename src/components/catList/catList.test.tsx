import React from 'react';
import {MemoryRouter} from 'react-router-dom';
import {render, screen} from '@testing-library/react';
import {userEvent} from '@testing-library/user-event';
import type {CatList as CatListType} from 'cat-lover/types/cats';
import * as UseFetchCats from '../../hooks/useFetchCats';
import {CatList} from './catList';

const mockUseFetchCats = jest.fn<
  ReturnType<typeof UseFetchCats.useFetchCats>,
  Parameters<typeof UseFetchCats.useFetchCats>
>();
jest.mock<typeof UseFetchCats>('../../hooks/useFetchCats', () => ({
  useFetchCats: (...args) => mockUseFetchCats(...args)
}));

describe('<CatList />', () => {
  describe('when fetching', () => {
    beforeEach(() => {
      mockUseFetchCats.mockReturnValue({
        isFetching: true,
        isReFetching: false,
        list: [],
        reFetch: jest.fn()
      });
    });

    it('should render loader', () => {
      const {container} = render(
        <MemoryRouter initialEntries={['/cats']}>
          <CatList />
        </MemoryRouter>
      );

      expect(container).toMatchSnapshot();
    });
  });

  describe('when fetching is complete', () => {
    const mockCats: CatListType = [
      {
        height: 0,
        id: '1',
        url: 'cat1.jpg',
        width: 0
      },
      {
        height: 0,
        id: '2',
        url: 'cat2.jpg',
        width: 0
      }
    ];

    beforeEach(() => {
      mockUseFetchCats.mockReturnValue({
        isFetching: false,
        isReFetching: false,
        list: mockCats,
        reFetch: jest.fn()
      });
    });

    it('should render a list of cats', () => {
      const {container} = render(
        <MemoryRouter initialEntries={['/cats']}>
          <CatList />
        </MemoryRouter>
      );

      expect(container).toMatchSnapshot();
    });
  });

  describe('when the user clicks on the "Load more" button', () => {
    const reFetchMock = jest.fn();

    beforeEach(() => {
      mockUseFetchCats.mockReturnValue({
        isFetching: false,
        isReFetching: false,
        list: [],
        reFetch: reFetchMock
      });
    });

    it('should re-fetch a list of cats', async () => {
      render(
        <MemoryRouter initialEntries={['/cats']}>
          <CatList />
        </MemoryRouter>
      );
      const loadMoreButton = screen.getByRole('button', {name: 'Load more'});

      await userEvent.click(loadMoreButton);

      expect(reFetchMock).toHaveBeenCalledTimes(1);
      expect(reFetchMock).toHaveBeenCalledWith();
    });

    describe('while re-fetching', () => {
      beforeEach(() => {
        mockUseFetchCats.mockReturnValue({
          isFetching: false,
          isReFetching: true,
          list: [],
          reFetch: jest.fn()
        });
      });

      it('should disable the "Load more" button', () => {
        render(
          <MemoryRouter initialEntries={['/cats']}>
            <CatList />
          </MemoryRouter>
        );
        const loadMoreButton = screen.getByRole('button', {name: 'Load more'});

        expect(loadMoreButton).toBeDisabled();
      });
    });
  });
});
