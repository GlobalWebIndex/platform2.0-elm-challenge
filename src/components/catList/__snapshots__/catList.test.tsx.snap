// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CatList /> when fetching is complete should render a list of cats 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="cats"
    >
      <a
        href="/1"
      >
        <img
          alt="Cat"
          class="image"
          src="cat1.jpg"
        />
      </a>
      <a
        href="/2"
      >
        <img
          alt="Cat"
          class="image"
          src="cat2.jpg"
        />
      </a>
    </div>
    <div
      class="loadMore"
    >
      <button
        class="loadMore"
      >
        Load more
      </button>
    </div>
  </div>
</div>
`;

exports[`<CatList /> when fetching should render loader 1`] = `
<div>
  <div
    class="loader"
  >
    <svg>
      <use
        href="function SvgIcons(props) {
  return React.createElement(
    'svg', 
    Object.assign({}, props, {'data-file-name': SvgIcons.name})
  );
}#icon-cw"
      />
    </svg>
  </div>
</div>
`;
